File: game_engine/moves.py
from typing import List, Tuple, Optional
import copy

from game_engine.board import Board
from game_engine.models import GameEventResult
from game_engine.endgame import is_game_over
from game_engine.message_codes import (
    TURN_NOW,
    MOVE_PASSED,
    CAPTURE_CONTINUE,
    CAPTURE_CONTINUE_SAME,
    CAPTURE_MUST,
    CAPTURE_MUST_CONTINUE,
    MOVE_UNKNOWN_PIECE,
    MOVE_NO_CAPTURE_TARGET,
    MOVE_TARGET_OCCUPIED,
    PIECE_PROMOTED,
    VALIDATION_TO_MESSAGE,
)
from game_engine.validation import (
    get_all_mandatory_captures,
    batyr_can_continue_capture,
    find_captured_enemy,
    validate_move_with_code,
)
from game_engine.dictionaries import shatra_and_biy_possible_captures

PROMOTION_FOR_WHITE = {1, 2, 3}
PROMOTION_FOR_BLACK = {60, 61, 62}


def _promote_shatra(cells: dict, cell: int, color: str) -> bool:
    cell_name = cells.get(cell)
    if not cell_name or "шатра" not in cell_name:
        return False
    if color == "белый" and cell in PROMOTION_FOR_WHITE:
        cells[cell] = "белый батыр"
        return True
    if color == "черный" and cell in PROMOTION_FOR_BLACK:
        cells[cell] = "черный батыр"
        return True
    return False


def execute_move(
    cells: dict,
    from_cell: int,
    to_cell: int,
    current_color: str,
    batyr_captured_this_turn: List[int] = None,
) -> Tuple[dict, List[int], List[int]]:
    if batyr_captured_this_turn is None:
        batyr_captured_this_turn = []

    new_cells = copy.deepcopy(cells)
    board = Board(new_cells)
    captured_positions = []
    new_batyr_captures = copy.copy(batyr_captured_this_turn)

    piece = board.get_piece_object(from_cell)

    if piece and piece.can_capture(cells, from_cell, to_cell, batyr_captured_this_turn):
        enemy_cell = find_captured_enemy(cells, piece, from_cell, to_cell, batyr_captured_this_turn)
        board.move_piece(from_cell, to_cell)
        if enemy_cell:
            board.remove_piece(enemy_cell)
            captured_positions.append(enemy_cell)
            if piece.get_type() == "батыр":
                new_batyr_captures.append(enemy_cell)
        return board.copy_cells(), captured_positions, new_batyr_captures

    board.move_piece(from_cell, to_cell)
    if piece and piece.get_type() == "батыр":
        new_batyr_captures.clear()
    return board.copy_cells(), captured_positions, new_batyr_captures


def _game_over_result(
    positions: dict,
    *,
    winner_color: str | None = None,
    draw_reason: str | None = None,
    captured_positions: List[int] | None = None,
    captured_pieces: List[int] | None = None,
) -> GameEventResult:
    return GameEventResult(
        movers_color=None,
        updated_positions=positions,
        game_over=True,
        winner_color=winner_color,
        draw_reason=draw_reason,
        captured_positions=captured_positions or [],
        captured_pieces=captured_pieces or [],
    )


def _validation_error(code: str) -> GameEventResult:
    message_code = VALIDATION_TO_MESSAGE.get(code, "move.illegal")
    return GameEventResult(message_code=message_code)


def process_move(
    cells: dict,
    current_color: str,
    from_cell: int,
    to_cell: int,
    chain_capture_cell: int = None,
    batyr_captured_this_turn: List[int] = None,
    position_history: dict = None,
    moves_with_two_biys: int = 0,
) -> GameEventResult:
    if batyr_captured_this_turn is None:
        batyr_captured_this_turn = []
    if position_history is None:
        position_history = {}

    current_batyr_captures = copy.copy(batyr_captured_this_turn)
    board_copy = copy.deepcopy(cells)

    board_obj = Board(board_copy)
    over, winner_color, draw_reason = is_game_over(board_obj, position_history, moves_with_two_biys)
    if over:
        return _game_over_result(cells, winner_color=winner_color, draw_reason=draw_reason)

    if chain_capture_cell == 0:
        over, winner_color, draw_reason = is_game_over(
            Board(board_copy), position_history, moves_with_two_biys,
        )
        return _finish_move(
            positions=board_copy,
            mover_color=current_color,
            message_code=MOVE_PASSED,
            history=True,
            clear_pending=True,
            game_over=over,
            winner_color=winner_color,
            draw_reason=draw_reason,
        )

    if chain_capture_cell is not None and chain_capture_cell != 0:
        return _process_chain_capture(
            board_copy, cells, current_color, from_cell, to_cell,
            chain_capture_cell, current_batyr_captures, position_history, moves_with_two_biys,
        )

    if chain_capture_cell is None or chain_capture_cell == 0:
        valid, _error_msg, code = validate_move_with_code(
            board_copy, from_cell, to_cell, current_color, current_batyr_captures,
        )
        if not valid:
            return _validation_error(code)

    new_cells, captured_positions, new_batyr_captures = execute_move(
        board_copy, from_cell, to_cell, current_color, current_batyr_captures,
    )

    piece = Board(board_copy).get_piece_object(from_cell)
    if piece is None:
        piece = Board(board_copy).get_piece_object(to_cell)
    piece_kind = piece.get_type() if piece else ""

    if piece_kind == "шатра":
        if _promote_shatra(new_cells, to_cell, current_color):
            if not captured_positions:
                return _finish_move(
                    positions=new_cells,
                    mover_color=current_color,
                    message_code=PIECE_PROMOTED,
                    message_params={"color": current_color},
                    history=True,
                    clear_pending=True,
                    captured_positions=captured_positions,
                )
            piece_kind = "батыр"

    next_board = Board(new_cells)
    over, winner_color, draw_reason = is_game_over(next_board, position_history, moves_with_two_biys)
    if over:
        return _game_over_result(
            new_cells,
            winner_color=winner_color,
            draw_reason=draw_reason,
            captured_positions=captured_positions,
        )

    has_captured = len(captured_positions) > 0
    can_continue_chain = False
    if has_captured:
        if piece_kind in ["шатра", "бий"]:
            piece_at_land = Board(new_cells).get_piece_object(to_cell)
            if piece_at_land:
                for to_cell_next in shatra_and_biy_possible_captures.get(to_cell, {}):
                    if piece_at_land.can_capture(new_cells, to_cell, to_cell_next, new_batyr_captures):
                        can_continue_chain = True
                        break
        else:
            can_continue_chain = batyr_can_continue_capture(
                next_board, to_cell, current_color, new_batyr_captures,
            )

    can_pass_turn = piece_kind == "бий" and has_captured

    if can_continue_chain:
        return GameEventResult(
            message_code=CAPTURE_CONTINUE,
            movers_color=current_color,
            updated_positions=new_cells,
            captured_positions=captured_positions,
            opportunity_pass_the_move=can_pass_turn,
            position_for_mandatory_capture=to_cell,
            captured_pieces=new_batyr_captures,
        )

    if has_captured and not can_continue_chain:
        next_player = _opponent(current_color)
        over, winner_color, draw_reason = is_game_over(next_board, position_history, moves_with_two_biys)
        return _finish_move(
            positions=new_cells,
            mover_color=current_color,
            message_code=TURN_NOW,
            message_params={"color": next_player},
            history=True,
            clear_pending=True,
            game_over=over,
            winner_color=winner_color,
            draw_reason=draw_reason,
            captured_positions=captured_positions,
            captured_pieces=new_batyr_captures,
            opportunity_pass=can_pass_turn,
        )

    next_player = _opponent(current_color)
    over, winner_color, draw_reason = is_game_over(next_board, position_history, moves_with_two_biys)

    chain_capture_pos = None
    if next_player and not over:
        mandatory_captures = get_all_mandatory_captures(Board(new_cells), next_player)
        if mandatory_captures:
            chain_capture_pos = mandatory_captures[0][0]

    return _finish_move(
        positions=new_cells,
        mover_color=current_color,
        message_code=TURN_NOW,
        message_params={"color": next_player},
        history=True,
        clear_pending=True,
        game_over=over,
        winner_color=winner_color,
        draw_reason=draw_reason,
        captured_positions=captured_positions,
        opportunity_pass=can_pass_turn,
        mandatory_pos=chain_capture_pos,
    )


def _process_chain_capture(
    board_copy, cells, current_color, from_cell, to_cell,
    chain_capture_cell, current_batyr_captures, position_history=None, moves_with_two_biys=0,
) -> GameEventResult:
    if from_cell != chain_capture_cell:
        return GameEventResult(
            message_code=CAPTURE_CONTINUE_SAME,
            movers_color=current_color,
            updated_positions=cells,
        )
    piece = Board(board_copy).get_piece_object(from_cell)
    if piece and piece.get_type() in ["шатра", "бий"]:
        return _process_chain_shatra_biy(
            board_copy, cells, current_color, from_cell, to_cell,
            current_batyr_captures, piece, position_history, moves_with_two_biys,
        )
    if piece and piece.get_type() == "батыр":
        return _process_chain_batyr(
            board_copy, cells, current_color, from_cell, to_cell,
            current_batyr_captures, piece, position_history, moves_with_two_biys,
        )
    return GameEventResult(
        message_code=MOVE_UNKNOWN_PIECE,
        movers_color=current_color,
        updated_positions=cells,
    )


def _process_chain_shatra_biy(
    board_copy, cells, current_color, from_cell, to_cell,
    current_batyr_captures, piece, position_history=None, moves_with_two_biys=0,
) -> GameEventResult:
    if not piece.can_capture(board_copy, from_cell, to_cell, current_batyr_captures):
        return GameEventResult(
            message_code=CAPTURE_MUST,
            movers_color=current_color,
            updated_positions=cells,
        )

    new_cells, captured_positions, new_batyr_captures = execute_move(
        board_copy, from_cell, to_cell, current_color, current_batyr_captures,
    )
    _promote_shatra(new_cells, to_cell, current_color)
    board = Board(new_cells)
    piece_kind = piece.get_type()

    over, winner_color, draw_reason = is_game_over(board, position_history, moves_with_two_biys)
    if over:
        return _game_over_result(
            new_cells,
            winner_color=winner_color,
            draw_reason=draw_reason,
            captured_positions=captured_positions,
            captured_pieces=new_batyr_captures,
        )

    can_continue_chain = False
    piece_at_land = board.get_piece_object(to_cell)
    if piece_at_land:
        if piece_at_land.get_type() == "батыр":
            can_continue_chain = batyr_can_continue_capture(
                board, to_cell, current_color, new_batyr_captures,
            )
        else:
            for to_cell_next in shatra_and_biy_possible_captures.get(to_cell, {}):
                if piece_at_land.can_capture(new_cells, to_cell, to_cell_next, new_batyr_captures):
                    can_continue_chain = True
                    break

    can_pass_turn = piece_kind == "бий"

    if can_continue_chain:
        return GameEventResult(
            message_code=CAPTURE_CONTINUE,
            movers_color=current_color,
            updated_positions=new_cells,
            captured_positions=captured_positions,
            opportunity_pass_the_move=can_pass_turn,
            position_for_mandatory_capture=to_cell,
            captured_pieces=new_batyr_captures,
        )

    next_player = _opponent(current_color)
    return _finish_move(
        positions=new_cells,
        mover_color=current_color,
        message_code=TURN_NOW,
        message_params={"color": next_player},
        history=True,
        clear_pending=True,
        captured_positions=captured_positions,
        opportunity_pass=can_pass_turn,
    )


def _process_chain_batyr(
    board_copy, cells, current_color, from_cell, to_cell,
    current_batyr_captures, piece, position_history=None, moves_with_two_biys=0,
) -> GameEventResult:
    if not piece.can_capture(board_copy, from_cell, to_cell, current_batyr_captures):
        return GameEventResult(
            message_code=CAPTURE_MUST_CONTINUE,
            movers_color=current_color,
            updated_positions=cells,
        )

    new_cells, captured_positions, new_batyr_captures = execute_move(
        board_copy, from_cell, to_cell, current_color, current_batyr_captures,
    )
    board = Board(new_cells)

    over, winner_color, draw_reason = is_game_over(board, position_history, moves_with_two_biys)
    if over:
        return _game_over_result(
            new_cells,
            winner_color=winner_color,
            draw_reason=draw_reason,
            captured_positions=captured_positions,
        )

    can_continue = batyr_can_continue_capture(board, to_cell, current_color, new_batyr_captures)

    if can_continue:
        return GameEventResult(
            message_code=CAPTURE_CONTINUE,
            movers_color=current_color,
            updated_positions=new_cells,
            captured_positions=captured_positions,
            position_for_mandatory_capture=to_cell,
            captured_pieces=new_batyr_captures,
        )

    next_player = _opponent(current_color)
    return _finish_move(
        positions=new_cells,
        mover_color=current_color,
        message_code=TURN_NOW,
        message_params={"color": next_player},
        history=True,
        clear_pending=True,
        captured_positions=captured_positions,
        captured_pieces=new_batyr_captures,
    )


def _finish_move(
    positions: dict,
    mover_color: str,
    message_code: str = "",
    message_params: dict | None = None,
    history: bool = False,
    clear_pending: bool = False,
    game_over: bool = False,
    winner_color: str | None = None,
    draw_reason: str | None = None,
    captured_positions: List[int] | None = None,
    opportunity_pass: bool = False,
    mandatory_pos: int | None = None,
    captured_pieces: List[int] | None = None,
) -> GameEventResult:
    next_mover = _opponent(mover_color)
    return GameEventResult(
        message_code=message_code,
        message_params=message_params or {},
        movers_color=next_mover,
        updated_positions=positions,
        captured_positions=captured_positions or [],
        game_over=game_over,
        winner_color=winner_color,
        draw_reason=draw_reason,
        opportunity_pass_the_move=opportunity_pass,
        position_for_mandatory_capture=mandatory_pos,
        captured_pieces=captured_pieces or [],
    )


def _opponent(color: str) -> str:
    return "черный" if color == "белый" else "белый"


def has_mandatory_from_position(cells: dict, color: str, pos: int = None) -> bool:
    board = Board(cells)
    mandatory = get_all_mandatory_captures(board, color)
    if pos is not None:
        return any(f == pos for f, _ in mandatory)
    return len(mandatory) > 0

File: frontend/packages/shatra-rules/src/moves.js
import { Board } from './board.js';
import { shatraAndBiyPossibleCaptures } from './dictLoader.js';
import {
  validateMove,
  getAllMandatoryCaptures,
  findCapturedEnemy,
  batyrCanContinueCapture,
} from './hints.js';
import { isGameOver, recordPosition } from './endgame.js';
import {
  TURN_NOW,
  MOVE_PASSED,
  CAPTURE_CONTINUE,
  CAPTURE_CONTINUE_SAME,
  CAPTURE_MUST,
  CAPTURE_MUST_CONTINUE,
  MOVE_UNKNOWN_PIECE,
  PIECE_PROMOTED,
  VALIDATION_TO_MESSAGE,
} from './messageCodes.js';

const PROMOTION_FOR_WHITE = new Set([1, 2, 3]);
const PROMOTION_FOR_BLACK = new Set([60, 61, 62]);

function cloneCells(cells) {
  const out = {};
  for (const [k, v] of Object.entries(cells || {})) {
    out[Number(k)] = v ?? null;
  }
  return out;
}

function opponent(color) {
  return color === 'белый' ? 'черный' : 'белый';
}

function makeResult(fields = {}) {
  return {
    messageCode: '',
    messageParams: {},
    moversColor: null,
    updatedPositions: null,
    capturedPositions: [],
    gameOver: false,
    winnerColor: null,
    drawReason: null,
    opportunityPassTheMove: false,
    positionForMandatoryCapture: null,
    capturedPieces: [],
    ...fields,
  };
}

function validationError(code) {
  return makeResult({
    messageCode: VALIDATION_TO_MESSAGE[code] || 'move.illegal',
  });
}

function gameOverResult(positions, {
  winnerColor = null,
  drawReason = null,
  capturedPositions = [],
  capturedPieces = [],
} = {}) {
  return makeResult({
    moversColor: null,
    updatedPositions: positions,
    gameOver: true,
    winnerColor,
    drawReason,
    capturedPositions,
    capturedPieces,
  });
}

function checkGameEnd(cells, positionHistory, movesWithTwoBiys, { record = true } = {}) {
  if (record) recordPosition(positionHistory, cells);
  const { over, winnerColor, drawReason } = isGameOver(
    new Board(cells),
    positionHistory,
    movesWithTwoBiys,
  );
  return { over, winnerColor, drawReason };
}

function chainCaptureAfterTurn(board, nextPlayer) {
  const mandatoryCaptures = getAllMandatoryCaptures(board, nextPlayer);
  if (!mandatoryCaptures.length) return null;
  const hasNonBiyAttacker = mandatoryCaptures.some(([fromCell]) => {
    const piece = board.getPieceObject(fromCell);
    return piece && piece.getType() !== 'бий';
  });
  if (!hasNonBiyAttacker) return null;
  return mandatoryCaptures[0][0];
}

function promoteShatra(cells, cell, color) {
  const cellName = cells[cell];
  if (!cellName || !cellName.includes('шатра')) return false;
  if (color === 'белый' && PROMOTION_FOR_WHITE.has(cell)) {
    cells[cell] = 'белый батыр';
    return true;
  }
  if (color === 'черный' && PROMOTION_FOR_BLACK.has(cell)) {
    cells[cell] = 'черный батыр';
    return true;
  }
  return false;
}

function executeMove(cells, fromCell, toCell, currentColor, batyrCapturedThisTurn = []) {
  const caps = [...(batyrCapturedThisTurn || [])];
  const newCells = cloneCells(cells);
  const board = new Board(newCells);
  const capturedPositions = [];
  const newBatyrCaptures = [...caps];

  const piece = board.getPieceObject(fromCell);
  if (piece && piece.canCapture(cells, fromCell, toCell, batyrCapturedThisTurn)) {
    const enemyCell = findCapturedEnemy(cells, piece, fromCell, toCell, batyrCapturedThisTurn);
    board.movePiece(fromCell, toCell);
    if (enemyCell != null) {
      board.removePiece(enemyCell);
      capturedPositions.push(enemyCell);
      if (piece.getType() === 'батыр') {
        newBatyrCaptures.push(enemyCell);
      }
    }
    return [board.copyCells(), capturedPositions, newBatyrCaptures];
  }

  board.movePiece(fromCell, toCell);
  if (piece && piece.getType() === 'батыр') {
    newBatyrCaptures.length = 0;
  }
  return [board.copyCells(), capturedPositions, newBatyrCaptures];
}

function finishMove(positions, moverColor, {
  messageCode = '',
  messageParams = {},
  gameOver = false,
  winnerColor = null,
  drawReason = null,
  capturedPositions = [],
  opportunityPass = false,
  mandatoryPos = null,
  capturedPieces = [],
} = {}) {
  return makeResult({
    messageCode,
    messageParams,
    moversColor: opponent(moverColor),
    updatedPositions: positions,
    capturedPositions,
    gameOver,
    winnerColor,
    drawReason,
    opportunityPassTheMove: opportunityPass,
    positionForMandatoryCapture: mandatoryPos,
    capturedPieces,
  });
}

function processChainShatraBiy(
  boardCopy,
  cells,
  currentColor,
  fromCell,
  toCell,
  currentBatyrCaptures,
  piece,
  positionHistory,
  movesWithTwoBiys,
) {
  if (!piece.canCapture(boardCopy, fromCell, toCell, currentBatyrCaptures)) {
    return makeResult({
      messageCode: CAPTURE_MUST,
      moversColor: currentColor,
      updatedPositions: cells,
    });
  }

  const [newCells, capturedPositions, newBatyrCaptures] = executeMove(
    boardCopy,
    fromCell,
    toCell,
    currentColor,
    currentBatyrCaptures,
  );
  promoteShatra(newCells, toCell, currentColor);
  const board = new Board(newCells);
  const pieceKind = piece.getType();

  const { over, winnerColor, drawReason } = checkGameEnd(
    newCells,
    positionHistory,
    movesWithTwoBiys,
  );
  if (over) {
    return gameOverResult(newCells, {
      winnerColor,
      drawReason,
      capturedPositions,
      capturedPieces: newBatyrCaptures,
    });
  }

  let canContinueChain = false;
  const pieceAtLand = board.getPieceObject(toCell);
  if (pieceAtLand) {
    if (pieceAtLand.getType() === 'батыр') {
      canContinueChain = batyrCanContinueCapture(
        board,
        toCell,
        currentColor,
        newBatyrCaptures,
      );
    } else {
      for (const toCellNext of Object.keys(shatraAndBiyPossibleCaptures[toCell] || {})) {
        if (pieceAtLand.canCapture(newCells, toCell, Number(toCellNext), newBatyrCaptures)) {
          canContinueChain = true;
          break;
        }
      }
    }
  }

  const canPassTurn = pieceKind === 'бий' && canContinueChain;
  if (canContinueChain) {
    return makeResult({
      messageCode: CAPTURE_CONTINUE,
      moversColor: currentColor,
      updatedPositions: newCells,
      capturedPositions,
      opportunityPassTheMove: canPassTurn,
      positionForMandatoryCapture: toCell,
      capturedPieces: newBatyrCaptures,
    });
  }

  return finishMove(newCells, currentColor, {
    messageCode: TURN_NOW,
    messageParams: { color: opponent(currentColor) },
    capturedPositions,
    opportunityPass: canPassTurn,
  });
}

function processChainBatyr(
  boardCopy,
  cells,
  currentColor,
  fromCell,
  toCell,
  currentBatyrCaptures,
  piece,
  positionHistory,
  movesWithTwoBiys,
) {
  if (!piece.canCapture(boardCopy, fromCell, toCell, currentBatyrCaptures)) {
    return makeResult({
      messageCode: CAPTURE_MUST_CONTINUE,
      moversColor: currentColor,
      updatedPositions: cells,
    });
  }

  const [newCells, capturedPositions, newBatyrCaptures] = executeMove(
    boardCopy,
    fromCell,
    toCell,
    currentColor,
    currentBatyrCaptures,
  );
  const board = new Board(newCells);

  const { over, winnerColor, drawReason } = checkGameEnd(
    newCells,
    positionHistory,
    movesWithTwoBiys,
  );
  if (over) {
    return gameOverResult(newCells, {
      winnerColor,
      drawReason,
      capturedPositions,
    });
  }

  const canContinue = batyrCanContinueCapture(
    board,
    toCell,
    currentColor,
    newBatyrCaptures,
  );
  if (canContinue) {
    return makeResult({
      messageCode: CAPTURE_CONTINUE,
      moversColor: currentColor,
      updatedPositions: newCells,
      capturedPositions,
      positionForMandatoryCapture: toCell,
      capturedPieces: newBatyrCaptures,
    });
  }

  return finishMove(newCells, currentColor, {
    messageCode: TURN_NOW,
    messageParams: { color: opponent(currentColor) },
    capturedPositions,
    capturedPieces: newBatyrCaptures,
  });
}

function processChainCapture(
  boardCopy,
  cells,
  currentColor,
  fromCell,
  toCell,
  chainCaptureCell,
  currentBatyrCaptures,
  positionHistory,
  movesWithTwoBiys,
) {
  if (fromCell !== chainCaptureCell) {
    return makeResult({
      messageCode: CAPTURE_CONTINUE_SAME,
      moversColor: currentColor,
      updatedPositions: cells,
    });
  }

  const piece = new Board(boardCopy).getPieceObject(fromCell);
  if (piece && (piece.getType() === 'шатра' || piece.getType() === 'бий')) {
    return processChainShatraBiy(
      boardCopy,
      cells,
      currentColor,
      fromCell,
      toCell,
      currentBatyrCaptures,
      piece,
      positionHistory,
      movesWithTwoBiys,
    );
  }
  if (piece && piece.getType() === 'батыр') {
    return processChainBatyr(
      boardCopy,
      cells,
      currentColor,
      fromCell,
      toCell,
      currentBatyrCaptures,
      piece,
      positionHistory,
      movesWithTwoBiys,
    );
  }
  return makeResult({
    messageCode: MOVE_UNKNOWN_PIECE,
    moversColor: currentColor,
    updatedPositions: cells,
  });
}

/**
 * Apply a move locally (mirrors game_engine.moves.process_move).
 */
export function processMove({
  cells,
  currentColor,
  fromCell,
  toCell,
  chainCaptureCell = null,
  batyrCapturedThisTurn = [],
  positionHistory = {},
  movesWithTwoBiys = 0,
}) {
  const currentBatyrCaptures = [...(batyrCapturedThisTurn || [])];
  const boardCopy = cloneCells(cells);
  const history = positionHistory;

  const { over, winnerColor, drawReason } = isGameOver(
    new Board(boardCopy),
    history,
    movesWithTwoBiys,
  );
  if (over) {
    return gameOverResult(cells, { winnerColor, drawReason });
  }

  if (chainCaptureCell === 0) {
    const endCheck = isGameOver(new Board(boardCopy), history, movesWithTwoBiys);
    return finishMove(boardCopy, currentColor, {
      messageCode: MOVE_PASSED,
      gameOver: endCheck.over,
      winnerColor: endCheck.winnerColor,
      drawReason: endCheck.drawReason,
    });
  }

  if (chainCaptureCell != null && chainCaptureCell !== 0) {
    return processChainCapture(
      boardCopy,
      cells,
      currentColor,
      fromCell,
      toCell,
      chainCaptureCell,
      currentBatyrCaptures,
      history,
      movesWithTwoBiys,
    );
  }

  const { valid, code } = validateMove(
    boardCopy,
    fromCell,
    toCell,
    currentColor,
    currentBatyrCaptures,
  );
  if (!valid) {
    return validationError(code);
  }

  const [newCells, capturedPositions, newBatyrCaptures] = executeMove(
    boardCopy,
    fromCell,
    toCell,
    currentColor,
    currentBatyrCaptures,
  );

  let piece = new Board(boardCopy).getPieceObject(fromCell);
  if (!piece) piece = new Board(boardCopy).getPieceObject(toCell);
  let pieceKind = piece ? piece.getType() : '';

  if (pieceKind === 'шатра') {
    if (promoteShatra(newCells, toCell, currentColor)) {
      if (!capturedPositions.length) {
        return finishMove(newCells, currentColor, {
          messageCode: PIECE_PROMOTED,
          messageParams: { color: currentColor },
          capturedPositions,
        });
      }
      pieceKind = 'батыр';
    }
  }

  let end = checkGameEnd(newCells, history, movesWithTwoBiys);
  if (end.over) {
    return gameOverResult(newCells, {
      winnerColor: end.winnerColor,
      drawReason: end.drawReason,
      capturedPositions,
    });
  }

  const hasCaptured = capturedPositions.length > 0;
  let canContinueChain = false;
  if (hasCaptured) {
    if (pieceKind === 'шатра' || pieceKind === 'бий') {
      const pieceAtLand = new Board(newCells).getPieceObject(toCell);
      if (pieceAtLand) {
        for (const toCellNext of Object.keys(shatraAndBiyPossibleCaptures[toCell] || {})) {
          if (pieceAtLand.canCapture(newCells, toCell, Number(toCellNext), newBatyrCaptures)) {
            canContinueChain = true;
            break;
          }
        }
      }
    } else {
      canContinueChain = batyrCanContinueCapture(
        new Board(newCells),
        toCell,
        currentColor,
        newBatyrCaptures,
      );
    }
  }

  const canPassTurn = pieceKind === 'бий' && canContinueChain;
  if (canContinueChain) {
    return makeResult({
      messageCode: CAPTURE_CONTINUE,
      moversColor: currentColor,
      updatedPositions: newCells,
      capturedPositions,
      opportunityPassTheMove: canPassTurn,
      positionForMandatoryCapture: toCell,
      capturedPieces: newBatyrCaptures,
    });
  }

  if (hasCaptured && !canContinueChain) {
    const nextPlayer = opponent(currentColor);
    end = checkGameEnd(newCells, history, movesWithTwoBiys, { record: false });
    return finishMove(newCells, currentColor, {
      messageCode: TURN_NOW,
      messageParams: { color: nextPlayer },
      gameOver: end.over,
      winnerColor: end.winnerColor,
      drawReason: end.drawReason,
      capturedPositions,
      capturedPieces: newBatyrCaptures,
      opportunityPass: canPassTurn,
    });
  }

  const nextPlayer = opponent(currentColor);
  end = checkGameEnd(newCells, history, movesWithTwoBiys, { record: false });
  let chainCapturePos = null;
  if (nextPlayer && !end.over) {
    chainCapturePos = chainCaptureAfterTurn(new Board(newCells), nextPlayer);
  }

  return finishMove(newCells, currentColor, {
    messageCode: TURN_NOW,
    messageParams: { color: nextPlayer },
    gameOver: end.over,
    winnerColor: end.winnerColor,
    drawReason: end.drawReason,
    capturedPositions,
    opportunityPass: canPassTurn,
    mandatoryPos: chainCapturePos,
  });
}
